external-crates/move/crates/move-compiler/src/parser/ast.rs

```diff
@@ -650,6 +711,253 @@ impl LeadingNameAccess_ {
     }
 }
 
+impl NameAccessChain_ {
+    pub fn single(name: Name) -> Self {
+        NameAccessChain_::Single(PathEntry {
+            name,
+            tyargs: None,
+            is_macro: None,
+        })
+    }
+
+    pub fn path(root: RootPathEntry) -> NamePath {
+        NamePath {
+            root,
+            entries: vec![],
+        }
+    }
+}
+
+impl NamePath {
+    /// Destructively take the type arguments, if any.
+    pub(crate) fn take_tyargs(&mut self) -> Option<Spanned<Vec<Type>>> {
+        if self.root.tyargs.is_some() {
+            return std::mem::take(&mut self.root.tyargs);
+        }
+        for entry in self.entries.iter_mut() {
+            if entry.tyargs.is_some() {
+                return std::mem::take(&mut entry.tyargs);
+            }
+        }
+        None
+    }
+}
+
+// Possibly move this trait out of `ast.rs`?
+#[allow(clippy::len_without_is_empty)]
+pub trait NameAccess {
+    fn is_macro(&self) -> Option<&Loc>;
+    fn tyargs(&self) -> Option<&Spanned<Vec<Type>>>;
+
+    fn push_path_entry(
+        &mut self,
+        name: Name,
+        tyargs: Option<Spanned<Vec<Type>>>,
+        is_macro: Option<Loc>,
+    ) -> Vec<Diagnostic>;
+
+    fn has_tyargs(&self) -> bool;
+    fn tyargs_loc(&self) -> Option<Loc>;
+    fn has_tyargs_last(&self) -> bool;
+    fn len(&self) -> usize;
+}
+
+impl NameAccess for PathEntry {
+    fn is_macro(&self) -> Option<&Loc> {
+        self.is_macro.as_ref()
+    }
+
+    fn tyargs(&self) -> Option<&Spanned<Vec<Type>>> {
+        self.tyargs.as_ref()
+    }
+
+    fn push_path_entry(
+        &mut self,
+        name: Name,
+        _tyargs: Option<Spanned<Vec<Type>>>,
+        _is_macro: Option<Loc>,
+    ) -> Vec<Diagnostic> {
+        let diag = ice!((name.loc, "Tried adding this name to a Single chain"));
+        vec![diag]
+    }
+
+    fn has_tyargs(&self) -> bool {
+        self.tyargs.is_some()
+    }
+
+    fn has_tyargs_last(&self) -> bool {
+        true
+    }
+
+    fn tyargs_loc(&self) -> Option<Loc> {
+        self.tyargs.as_ref().map(|sp!(loc, _)| *loc)
+    }
+
+    fn len(&self) -> usize {
+        1
+    }
+}
+
+impl NameAccess for NamePath {
+    fn is_macro(&self) -> Option<&Loc> {
+        if let Some(loc) = &self.root.is_macro {
+            return Some(loc);
+        }
+        for entry in self.entries.iter() {
+            if let Some(loc) = &entry.is_macro {
+                return Some(loc);
+            }
+        }
+        None
```

Comment:
Not sure if its really better but....
```suggestion
        self.root.is_macro.as_ref().or_else(|| self.entries.iter().find_map(|e| e.is_macro.as_ref()))
```